in-progress/7228-unified-bbjs-interface.md

```diff
@@ -0,0 +1,247 @@
+
+Native implementation (`AVM` not supported)
+
+- `bb prove_output_all | prove_ultra_honk_output_all | prove_mega_honk_output_all | client_ivc_prove_output_all`
+
+Wasm implementation (`AVM, ClientIVC, MegaHonk` not supported)
+
+- `acirGetCircuitSizes`
+- `commonInitSlabAllocator`
+- `srsInitSrs`
+- `acirNewAcirComposer`
+- `acirInitProvingKey`
+- `acirCreateProof | acirProveUltraHonk`
+- `acirSerializeProofIntoFields | acirProofAsFieldsUltraHonk`
+- `acirGetVerificationKey`
+- `acirSerializeVerificationKeyIntoFields`
+
+:question: Do we really need this method? Is there a significant gain over calling `prove + getProofAsFields + getVerificationKey` on the bb CLI? We're using it today on bb-prover mostly.
+
+#### `getProofAsFields(proof: Buffer, numPublicInputs: number, flavor: Flavor): { proofAsFields: Fields }`
+
+Given a proof in raw binary format, returns it formatted as fields. Required for generating the recursive artifacts for recursive proving, along with the verification key as fields. Requires calling `prove` first. Note that we need to keep this method separate from `prove` since it could be different users who create the inner proof and who want to convert it to fields (see [Noir docs](https://noir-lang.org/docs/dev/how_to/how-to-recursion/#step-3---verification-and-proof-artifacts)).
+
+:question: Can we reimplement this logic on ts-land? We've received conflicting responses to this.
+
+Native implementation (does not support `ClientIVC, AVM`)
+
+- `bb proof_as_fields | proof_as_fields_honk` (assumes `_honk` supports both Ultra and Mega)
+
+Wasm implementation (does not support `ClientIVC, AVM, MegaHonk`)
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirSerializeProofIntoFields | acirProofAsFieldsUltraHonk`
+
+#### `getVerificationKey(circuit: Circuit, flavor: Flavor): { vk: Buffer, vkAsFields: Fields }`
+
+Computes the verification key for a given circuit. Returns the verification key in binary and as fields, which allows the client to read its hash and the number of public inputs for the circuit. Users should call this once per circuit they want to verify, and reuse it for verifying as many proofs as needed.
+
+:question: Can we avoid returning `vkAsFields`, and convert from binary to fields in ts-land?
+
+Native implementation (`ClientIVC, AVM` not supported)
+
+- `bb write_vk | write_vk_ultra_honk | write_vk_mega_honk`
+- `bb vk_as_fields | vk_as_fields_ultra_honk | vk_as_fields_mega_honk`
+
+Wasm implementation (only `UltraPlonk` supported)
+
+- `acirGetCircuitSizes`
+- `commonInitSlabAllocator(size)`
+- `srsInitSrs(size)`
+- `acirInitProvingKey`
+- `acirGetVerificationKey`
+- `acirSerializeVerificationKeyIntoFields`
+
+#### `verify(proof: Buffer, vk: Buffer, flavor: Flavor): { isValid: boolean }`
```

Comment:
We should also allow verification of recursion-friendly proofs as if I'm incorporating a proof generated from someone else into a proof of my own, I want to verify that the proof is correct before I generate my proof.

Reply:
Not sure I'm following: how would the interface for that look like? Isn't it just calling `verify` but with the recursion-friendly proof and a vk specific for it? Or is this related to aggregating one proof into another?

Reply:
Off the top of my head I'm not sure if the `vk` encodes this information, if so then no change is necessary.